core/authn.go

```diff
@@ -34,26 +36,54 @@ type tokenResult struct {
 
 func (a *apiAuthn) handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		err := a.auth(req)
-		if err != nil {
-			errorFormatter.Write(req.Context(), rw, err)
+		token, err0 := a.tokenAuth(req)
+		ctx := req.Context()
+		if err0 == nil && token != "" {
+			// if this request was successfully authenticated, pass the token along
+			ctx = context.WithValue(ctx, "token", token)
+		}
+
+		err1 := a.localhostAuth(req)
+		if err1 == nil {
+			ctx = context.WithValue(ctx, "localhost", true)
+		}
+
+		// TODO(tessr): move this to authz as part of ACL work
+		if err0 != nil {
+			errorFormatter.Write(ctx, rw, err0)
 			return
 		}
-		next.ServeHTTP(rw, req)
+
+		next.ServeHTTP(rw, req.WithContext(ctx))
 	})
 }
 
-func (a *apiAuthn) auth(req *http.Request) error {
+func (a *apiAuthn) localhostAuth(req *http.Request) error {
+	h, _, err := net.SplitHostPort(req.RemoteAddr)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if !net.ParseIP(h).IsLoopback() {
+		return errNotAuthenticated
+	}
+	return nil
+}
+
+func (a *apiAuthn) tokenAuth(req *http.Request) (string, error) {
 	user, pw, ok := req.BasicAuth()
+	// TODO(tessr): remove the following clause once ACLs are in place
 	if !ok && a.alt != nil && a.alt(req) {
-		return nil
+		return "", nil
 	}
 
+	// Is this the way we want to encode the token to be passed around? Or just user?
```

Comment:
So there's a couple schools of thought here. One is what you're talking about, a.k.a. [claims-based identity](https://en.wikipedia.org/wiki/Claims-based_identity).

But there's also an entirely different school of thought around authorization-centric credentials which include a sort of floating policy in the credential themselves.